File: leetcode/valid-parentheses.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
 public:
    bool isValid(string s)
    {
        stack<char> st;
        set<char> opening = {'(', '[', '{'};
        unordered_map<char, char> ending = {
            {')', '('},
            {']', '['},
            {'}', '{'},
        };

        for (auto c : s) {
            if (opening.find(c) != opening.end()) {
                st.push(c);
            } else {
                if (st.empty() || st.top() != ending[c])
                    return false;
                st.pop();
            }
        }
        return st.empty();
    }
};

File: leetcode/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        li = []

        h = {")": "(", "}": "{", "]": "["}

        for c in s:
            if c in h.keys():
                if not li or li[-1] != h[c]:
                    return False
                li.pop()
            else:
                li.append(c)

        return len(li) == 0
